version: "3"

services:
  mysql:
    image: mysql:8.0.26
    container_name: mysql
    command: --default-authentication-plugin=mysql_native_password
    environment:
      MYSQL_DATABASE: "chat_app"
      MYSQL_ROOT_PASSWORD: "rootpw"
    networks:
      - chat_app
    ports:
      - "3306:3306"
    deploy:
      restart_policy:
        condition: on-failure
        max_attempts: 3
    healthcheck:
      test: ["CMD", "mysqladmin" ,"ping", "-h", "localhost"]
      timeout: 20s
      retries: 10

  phpmyadmin:
    container_name: phpmyadmin
    depends_on:
      mysql:
        condition: service_healthy
    image: phpmyadmin/phpmyadmin
    networks:
      - chat_app
    ports:
      - "8081:80"
    environment:
      PMA_HOST: mysql
      MYSQL_ROOT_PASSWORD: rootpw
    deploy:
      restart_policy:
        condition: on-failure
        max_attempts: 3

  chat_app_backend:
    container_name: chat_app_backend
    depends_on:
      mysql:
        condition: service_healthy
    build:
      context: ./
      dockerfile: Docker/backend/Dockerfile
    command: mvn spring-boot:run
    ports:
      - "9000:9000"
    networks:
      - chat_app
    volumes:
      - ./backend:/backend
      - /backend/target
    deploy:
      restart_policy:
        condition: on-failure
        max_attempts: 3

  chat_app_frontend:
    container_name: chat_app_frontend
    build:
      dockerfile: Docker/frontend/Dockerfile
      context: ./
    command: yarn dev
    depends_on:
      - chat_app_backend
    networks:
      - chat_app
    volumes:
      - ./frontend:/frontend
      - /frontend/node_modules
    ports:
      - "3000:3000"
    deploy:
      restart_policy:
        condition: on-failure
        max_attempts: 5

networks:
  chat_app:
    external: true